Lib/_pyrepl/utils.py

```diff
@@ -41,25 +85,82 @@ def unbracket(s: str, including_content: bool = False) -> str:
     return s.translate(ZERO_WIDTH_TRANS)
 
 
-def disp_str(buffer: str) -> tuple[CharBuffer, CharWidths]:
-    r"""Decompose the input buffer into a printable variant.
+def gen_colors(buffer: str) -> Iterator[ColorSpan]:
+    """Returns a list of index spans to color using the given color tag.
+
+    The input `buffer` should be a valid start of a Python code block, i.e.
+    it cannot be a block starting in the middle of a multiline string.
+    """
+    for match in COLORIZE_RE.finditer(buffer):
+        yield from gen_color_spans(match)
+
+
+def gen_color_spans(re_match: Match[str]) -> Iterator[ColorSpan]:
+    """Generate non-empty color spans."""
+    for tag, data in re_match.groupdict().items():
+        if not data:
+            continue
+        span = Span.from_re(re_match, tag)
+        tag = COLORIZE_GROUP_NAME_MAP.get(tag, tag)
+        yield ColorSpan(span, cast(ColorTag, tag))
+        if data in IDENTIFIERS_AFTER:
+            if name_match := IDENTIFIER_RE.match(re_match.string, span.end + 1):
+                span = Span.from_re(name_match, 1)
+                yield ColorSpan(span, "DEFINITION")
+
+
+def disp_str(
+    buffer: str, colors: list[ColorSpan] | None = None, start_index: int = 0
+) -> tuple[CharBuffer, CharWidths]:
+    r"""Decompose the input buffer into a printable variant with applied colors.
 
     Returns a tuple of two lists:
-    - the first list is the input buffer, character by character;
+    - the first list is the input buffer, character by character, with color
+      escape codes added (while those codes contain multiple ASCII characters,
+      each code is considered atomic *and is attached for the corresponding
+      visible character*);
     - the second list is the visible width of each character in the input
       buffer.
 
+    Note on colors:
+    - The `colors` list, if provided, is partially consumed within. We're using
+      a list and not a generator since we need to hold onto the current
+      unfinished span between calls to disp_str in case of multiline strings.
+    - The `colors` list is computed from the start of the input block. `buffer`
+      is only a subset of that input block, a single line within. This is why
+      we need `start_index` to inform us which position is the start of `buffer`
+      actually within user input. This allows us to match color spans correctly.
+
     Examples:
     >>> utils.disp_str("a = 9")
     (['a', ' ', '=', ' ', '9'], [1, 1, 1, 1, 1])
+
+    >>> line = "while 1:"
+    >>> colors = list(utils.gen_colors(line))
+    >>> utils.disp_str(line, colors=colors)
+    (['\x1b[1;34mw', 'h', 'i', 'l', 'e\x1b[0m', ' ', '1', ':'], [1, 1, 1, 1, 1, 1, 1, 1])
+
     """
     chars: CharBuffer = []
     char_widths: CharWidths = []
 
     if not buffer:
         return chars, char_widths
 
-    for c in buffer:
+    while colors and colors[0].span.end < start_index:
+        # move past irrelevant spans
+        colors.pop(0)
```

Comment:
I know this list is not big enough that it matters but it would recommend modifying this to figure out the cutting index and then slicing once